awsvpc configuration for ECS

willthames opened the pull request:
##### SUMMARY
Enable awsvpc network mode for ecs task definitions, tasks and services

##### ISSUE TYPE
 - Feature Pull Request


##### COMPONENT NAME
ecs_taskdefinition

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 0a3da471f5) last updated 2018/01/02 10:31:55 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Nov  2 2017, 18:42:05) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

##### ADDITIONAL INFORMATION

mattclay commented on test/integration/host_vars/aws_local.yml:
```diff
@@ -0,0 +1,2 @@
+ansible_connection: local
+ansible_python_interpreter: "{{ ansible_playbook_python }}"
```
Files specific to a test target should be under that target's directory tree, not directly under `test/integration/`.

mattclay commented on test/integration/inventory:
```diff
@@ -6,6 +6,7 @@ testhost3 ansible_ssh_host=127.0.0.3
 testhost4 ansible_ssh_host=127.0.0.4
 # For testing fact gathering
 facthost[0:25] ansible_host=127.0.0.1 ansible_connection=local
+aws_local
```
Files specific to a test target should be under that target's directory tree, not directly under `test/integration/`.

mattclay commented on test/integration/targets/ecs_cluster/runme.sh:
```diff
@@ -0,0 +1,27 @@
+# This is needed for the ubuntu1604py3 tests
+# Ubuntu patches virtualenv to make the default python2
+# but for the python3 tests we need virtualenv to use python3
+PYTHON=$("python${ANSIBLE_TEST_PYTHON_VERSION:-}" -c "import sys; print(sys.executable)")
```
You should be able to use `$ANSIBLE_TEST_PYTHON_INTERPRETER` instead. I don't recall why our existing tests use `$ANSIBLE_TEST_PYTHON_VERSION` to find the interpreter.

willthames commented:
ready_for_review

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -265,26 +285,41 @@
+            self.module.fail_json(msg="Network configuration must include subnets")
+        if 'security_groups' in network_config:
+            groups = network_config['security_groups']
+            if any(not sg.startwith('sg-') for sg in groups):
```
s/startwith/startswith

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -265,26 +285,41 @@
+            groups = network_config['security_groups']
+            if any(not sg.startwith('sg-') for sg in groups):
+                try:
+                    vpc_id = self.ec2.describe_subnets(SubnetIds=result['subnets'][0])['Subnets'][0]['VpcId']
```
SubnetIds expects a list `[result['subnets'][0]]`

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -265,26 +285,41 @@
-
         region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
         self.ecs = boto3_conn(module, conn_type='client', resource='ecs', region=region, endpoint=ec2_url, **aws_connect_kwargs)
 
```
self.ec2 still needs to be defined here. Also, AnsibleAWSModule lets you do `self.ecs = module.client('ecs')` and `self.ec2 = module.client('ec2')`, so you could remove the imports get_aws_connection_info and boto3_conn.

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -334,21 +369,49 @@ def create_service(self, service_name, cluster_name, task_definition, load_balan
+        # There doesn't seem to be a nice way to inspect botocore to look
+        # for attributes (and networkConfiguration is not an explicit argument
+        # to e.g. ecs.run_task, it's just passed as a keyword argument)
+        return LooseVersion(botocore.__version__) >= LooseVersion('1.7.44')
```
You could inspect arguments a method takes but what you've done seems a lot more clear.
```
from botocore import ArgumentGenerator
return 'networkConfiguration' in dict(ArgumentGenerator().generate_skeleton(self.ecs._service_model.operation_model('RunTask').input_shape))
```

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_task.py:
```diff
@@ -168,6 +180,24 @@ def __init__(self, module):
+            self.module.fail_json(msg="Network configuration must include subnets")
+        if 'security_groups' in network_config:
+            groups = network_config['security_groups']
+            if any(not sg.startwith('sg-') for sg in groups):
```
needs to be `startswith` here too

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_task.py:
```diff
@@ -168,6 +180,24 @@ def __init__(self, module):
+            groups = network_config['security_groups']
+            if any(not sg.startwith('sg-') for sg in groups):
+                try:
+                    vpc_id = self.ec2.describe_subnets(SubnetIds=result['subnets'][0])['Subnets'][0]['VpcId']
```
SubnetIds needs to be a list here too

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -334,21 +369,49 @@ def create_service(self, service_name, cluster_name, task_definition, load_balan
+    def jsonize(self, service):
+        # some fields are datetime which is not JSON serializable
+        # make them strings
+        if 'deployments' in service:
```
There is a 'createdAt' as a key in service as well.

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -166,6 +167,10 @@ def register_task(self, family, task_role_arn, network_mode, container_definitio
                         if port in port_mapping:
                             port_mapping[port] = int(port_mapping[port])
+                    if network_mode == 'awsvpc' and 'hostPort' in port_mapping:
+                        if port_mapping['hostPort'] != port_mapping.get('containerPort'):
```
Might want to cast port_mapping['hostPort'] to an int just in case they passed a string. At some point we could probably update this module to use argspec suboptions so we could ensure the ports are integers before the module uses them.

willthames replied:
That's actually done in the three preceding lines :)

s-hertel replied:
Oops. :)

s-hertel commented on test/integration/targets/ecs_cluster/playbooks/full_test.yml:
```diff
@@ -0,0 +1,5 @@
+  connection: local
+
+  roles:
+    - ecs_cluster
```
This isn't something you changed, but do you mind changing this line to 'yes'? https://github.com/ansible/ansible/blob/devel/test/integration/targets/ecs_cluster/tasks/main.yml#L20
(if you don't have the correct instance role already the tests don't work because of that)

willthames replied:
ah right, thanks

willthames commented:
Thanks @s-hertel - there are definitely coverage issues with my tests - I've already found some of those  thanks to your review so that's a great start.

Commit:
Add VPC configuration to ECS modules

Enable awsvpc network mode for ECS services and tasks and
their underlying task definitions

Improve test suite to thoroughly test the changes

Use runme.sh technique to run old and new versions of botocore to
ensure that the modules work with older botocore and older network modes
and fail gracefully if awsvpc network mode is used with older botocore

willthames commented:
ready_for_review

ryansb commented:
needs_revision

willthames commented:
@ryansb would you mind elaborating on that feedback?

ryansb commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -334,21 +369,51 @@ def create_service(self, service_name, cluster_name, task_definition, load_balan
+
 
 def main():
     argument_spec = ec2_argument_spec()
```
ec2_argument_spec isn't needed anymore (built in to AnsibleAWSModule).

ryansb commented on lib/ansible/modules/cloud/amazon/ecs_task.py:
```diff
@@ -205,14 +234,26 @@ def start_task(self, cluster, task_definition, overrides, container_instances, s
+
 
 def main():
     argument_spec = ec2_argument_spec()
```
Same here - AnsibleAWSModule adds this itself.

ryansb reviewed (approved):
Would like those couple AWSModule tweaks, if you have time for a follow-up PR.

ryansb commented:
@willthames once you cover Matt's feedback on the integration test tree, this PR should be set to go.

mattclay reviewed (approved):
Looks OK.

NOTE: I haven't run the tests and they're not supported by CI.
